File: src/Database/Tests/PropertyTest.cpp
/**
 * @file
 * @copyright This code is licensed under the 3-clause BSD license.\n
 *            Copyright ETH Zurich, Department of Chemistry and Applied Biosciences, Reiher Group.\n
 *            See LICENSE.txt for details.
 */
#include <Database/Exceptions.h>
#include <Database/Manager.h>
#include <Database/Objects/Model.h>
#include <Database/Objects/NumberProperty.h>
#include <Database/Objects/Property.h>
#include <Database/Objects/VectorProperty.h>
#include <gmock/gmock.h>

using namespace testing;
namespace Scine {
namespace Database {
namespace Tests {
/**
 * @class PropertyTest
 * @brief Comprises tests for the class Scine::Database::Property.
 * @test
 */
struct PropertyTest : public Test {
  Credentials credentials{TEST_MONGO_DB_IP, std::atoi(TEST_MONGO_DB_PORT), "unittest_db_PropertyTest"};
  Manager db;

  void SetUp() final {
    db.setCredentials(credentials);
    db.connect();
    db.init();
  }
  void TearDown() final {
    db.wipe();
  }
};

TEST_F(PropertyTest, PropertyName) {
  // Setup
  auto coll = db.getCollection("properties");
  NumberProperty test = NumberProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), 0.0, coll);

  Property property(test.id(), coll);
  ASSERT_TRUE(property.hasId());

  // PropertyName Functionalities
  ASSERT_EQ(property.getPropertyName(), "density_matrix");
  property.setPropertyName("electronic_energy");
  ASSERT_EQ(property.getPropertyName(), "electronic_energy");
}

TEST_F(PropertyTest, PropertyNameFailsCollection) {
  Property property;
  ASSERT_THROW(property.getPropertyName(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.setPropertyName("asdf"), Exceptions::MissingLinkedCollectionException);
}

TEST_F(PropertyTest, PropertyNameFailsID) {
  auto coll = db.getCollection("properties");
  Property property;
  property.link(coll);
  ASSERT_THROW(property.getPropertyName(), Exceptions::MissingIDException);
  ASSERT_THROW(property.setPropertyName("asdf"), Exceptions::MissingIDException);
}

TEST_F(PropertyTest, Model) {
  // Setup
  auto coll = db.getCollection("properties");

  auto property = NumberProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), 0.0, coll);
  ASSERT_TRUE(property.hasId());

  // Model Functionalities
  auto model = property.getModel();
  ASSERT_EQ(model.method, "pbe");
  ASSERT_EQ(model.methodFamily, "dft");
  ASSERT_EQ(model.basisSet, "def2-svp");
  ASSERT_EQ(model.spinMode, "any");
  model.spinMode = "a";
  model.basisSet = "b";
  model.spinMode = "c";
  model.program = "d";
  model.version = "e";
  model.solvation = "f";
  model.solvent = "g";
  model.embedding = "h";
  model.periodicBoundaries = "i";
  model.externalField = "j";
  model.methodFamily = "k";
  property.setModel(model);
  auto model_db = property.getModel();
  ASSERT_EQ(model.spinMode, model_db.spinMode);
  ASSERT_EQ(model.basisSet, model_db.basisSet);
  ASSERT_EQ(model.spinMode, model_db.spinMode);
  ASSERT_EQ(model.program, model_db.program);
  ASSERT_EQ(model.version, model_db.version);
  ASSERT_EQ(model.solvation, model_db.solvation);
  ASSERT_EQ(model.solvent, model_db.solvent);
  ASSERT_EQ(model.embedding, model_db.embedding);
  ASSERT_EQ(model.periodicBoundaries, model_db.periodicBoundaries);
  ASSERT_EQ(model.externalField, model_db.externalField);
  ASSERT_EQ(model.methodFamily, model_db.methodFamily);
}

TEST_F(PropertyTest, ModelFailsCollection) {
  Property property;
  ASSERT_THROW(property.getModel(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.setModel(Model("dft", "pbe", "def2-svp")), Exceptions::MissingLinkedCollectionException);
}

TEST_F(PropertyTest, ModelFailsID) {
  auto coll = db.getCollection("properties");
  Property property;
  property.link(coll);
  ASSERT_THROW(property.getModel(), Exceptions::MissingIDException);
  ASSERT_THROW(property.setModel(Model("dft", "pbe", "def2-svp")), Exceptions::MissingIDException);
}

TEST_F(PropertyTest, Structure) {
  // Setup
  auto coll = db.getCollection("properties");

  auto property = NumberProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), 0.0, coll);
  ASSERT_TRUE(property.hasId());

  // tests
  ID id;
  ASSERT_FALSE(property.hasStructure());
  property.setStructure(id);
  ASSERT_TRUE(property.hasStructure());
  ASSERT_EQ(property.getStructure(), id);
  property.clearStructure();
  ASSERT_FALSE(property.hasStructure());
}

TEST_F(PropertyTest, StructureFailsCollection) {
  Property property;
  ID id;
  ASSERT_THROW(property.hasStructure(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.setStructure(id), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.getStructure(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.clearStructure(), Exceptions::MissingLinkedCollectionException);
}

TEST_F(PropertyTest, StructureFailsID) {
  auto coll = db.getCollection("properties");
  Property property;
  property.link(coll);
  ID id;
  ASSERT_THROW(property.hasStructure(), Exceptions::MissingIDException);
  ASSERT_THROW(property.setStructure(id), Exceptions::MissingIDException);
  ASSERT_THROW(property.getStructure(), Exceptions::MissingIDException);
  ASSERT_THROW(property.clearStructure(), Exceptions::MissingIDException);
}

TEST_F(PropertyTest, Calculation) {
  // Setup
  auto coll = db.getCollection("properties");
  auto property = NumberProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), 0.0, coll);
  ASSERT_TRUE(property.hasId());

  // tests
  ID id;
  ASSERT_FALSE(property.hasCalculation());
  property.setCalculation(id);
  ASSERT_TRUE(property.hasCalculation());
  ASSERT_EQ(property.getCalculation(), id);
  property.clearCalculation();
  ASSERT_FALSE(property.hasCalculation());
}

TEST_F(PropertyTest, CalculationFailsCollection) {
  Property property;
  ID id;
  ASSERT_THROW(property.getCalculation(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.setCalculation(id), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.hasCalculation(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.clearCalculation(), Exceptions::MissingLinkedCollectionException);
}

TEST_F(PropertyTest, CalculationFailsID) {
  auto coll = db.getCollection("properties");
  Property property;
  property.link(coll);
  ID id;
  ASSERT_THROW(property.getCalculation(), Exceptions::MissingIDException);
  ASSERT_THROW(property.setCalculation(id), Exceptions::MissingIDException);
  ASSERT_THROW(property.hasCalculation(), Exceptions::MissingIDException);
  ASSERT_THROW(property.clearCalculation(), Exceptions::MissingIDException);
}

TEST_F(PropertyTest, Comment) {
  // Setup
  auto coll = db.getCollection("properties");
  auto property = NumberProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), 0.0, coll);
  ASSERT_TRUE(property.hasId());

  // Comment Functionalities
  ASSERT_FALSE(property.hasComment());
  auto comment = property.getComment();
  ASSERT_EQ(comment, "");

  property.setComment("wubba lubba dub dub");
  ASSERT_TRUE(property.hasComment());
  comment = property.getComment();
  ASSERT_EQ(comment, "wubba lubba dub dub");

  property.clearComment();
  ASSERT_FALSE(property.hasComment());
  comment = property.getComment();
  ASSERT_EQ(comment, "");
}

TEST_F(PropertyTest, CommentFailsCollection) {
  Property property;
  ASSERT_THROW(property.hasComment(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.getComment(), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.setComment("foo"), Exceptions::MissingLinkedCollectionException);
  ASSERT_THROW(property.clearComment(), Exceptions::MissingLinkedCollectionException);
}

TEST_F(PropertyTest, CommentFailsID) {
  auto coll = db.getCollection("properties");
  Property property;
  property.link(coll);
  ASSERT_THROW(property.hasComment(), Exceptions::MissingIDException);
  ASSERT_THROW(property.getComment(), Exceptions::MissingIDException);
  ASSERT_THROW(property.setComment("foo"), Exceptions::MissingIDException);
  ASSERT_THROW(property.clearComment(), Exceptions::MissingIDException);
}

TEST_F(PropertyTest, IsOfType) {
  // Setup
  auto coll = db.getCollection("properties");
  auto property = NumberProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), 0.0, coll);
  property.link(coll);
  ASSERT_TRUE(property.hasId());
  // Test
  ASSERT_TRUE(property.isOfType<NumberProperty>());
  ASSERT_FALSE(property.isOfType<VectorProperty>());
}

TEST_F(PropertyTest, GetDerived) {
  // Setup
  auto coll = db.getCollection("properties");
  auto numberProperty = NumberProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), 5.6, coll);
  Property property(numberProperty.id(), coll);
  ASSERT_TRUE(property.hasId());
  // Test
  auto numprop = property.getDerived<NumberProperty>();
  ASSERT_EQ(numprop.getData(), 5.6);
}

TEST_F(PropertyTest, GetDerivedFails) {
  // Setup
  auto coll = db.getCollection("properties");
  auto test = NumberProperty::create("density_matrix", Model("dft", "pbe", "def2-svp"), 0.0, coll);
  Property property(test.id());
  ASSERT_THROW(property.getDerived<VectorProperty>(), Exceptions::MissingLinkedCollectionException);
  property.link(coll);
  ASSERT_TRUE(property.hasId());
  ASSERT_THROW(property.getDerived<VectorProperty>(), Exceptions::ObjectTypeMismatchException);
  Property property2;
  property2.link(coll);
  ASSERT_THROW(property2.getDerived<VectorProperty>(), Exceptions::MissingIDException);
}

} /* namespace Tests */
} /* namespace Database */
} /* namespace Scine */

File: src/Database/Python/Tests/property_test.py
# -*- coding: utf-8 -*-
__copyright__ = """This code is licensed under the 3-clause BSD license.
Copyright ETH Zurich, Department of Chemistry and Applied Biosciences, Reiher Group.
See LICENSE.txt for details.
"""

import scine_database as db
import unittest
import os


class PropertyTest(unittest.TestCase):
    def setUp(self):
        self.manager = db.Manager()
        self.manager.credentials.hostname = os.environ.get(
            'TEST_MONGO_DB_IP', '127.0.0.1')
        self.manager.credentials.port = int(os.environ.get('TEST_MONGO_DB_PORT', 27017))
        self.manager.credentials.database_name = "unittest_db_PropertyTest"
        self.manager.connect()
        self.manager.init()

    def tearDown(self):
        self.manager.wipe()

    def test_property_name(self):
        # Setup
        coll = self.manager.get_collection("properties")
        model = db.Model("dft", "pbe", "def2-svp")
        test = db.NumberProperty.make("density_matrix", model, 0.0, coll)
        prop = db.Property(test.id(), coll)
        assert prop.has_id()

        # PropertyName Functionalities
        assert prop.get_property_name() == "density_matrix"
        prop.set_property_name("electronic_energy")
        assert prop.get_property_name() == "electronic_energy"

    def test_property_name_fails_collection(self):
        prop = db.Property(db.ID())
        self.assertRaises(RuntimeError, lambda: prop.set_property_name("asdf"))
        self.assertRaises(RuntimeError, lambda: prop.get_property_name())

    def test_property_name_fails_id(self):
        coll = self.manager.get_collection("properties")
        prop = db.NumberProperty()
        prop.link(coll)
        self.assertRaises(RuntimeError, lambda: prop.set_property_name("asdf"))
        self.assertRaises(RuntimeError, lambda: prop.get_property_name())

    def test_model(self):
        # Setup
        coll = self.manager.get_collection("properties")

        model = db.Model("dft", "pbe", "def2-svp")
        prop = db.NumberProperty.make("density_matrix", model, 0.0, coll)
        assert prop.has_id()

        # Model Functionalities
        model = prop.get_model()
        assert model.method == "pbe"
        assert model.method_family == "dft"
        assert model.basis_set == "def2-svp"
        assert model.spin_mode == "any"
        model.basis_set = "b"
        model.spin_mode = "c"
        model.program = "d"
        model.version = "e"
        model.solvation = "f"
        model.solvent = "g"
        model.embedding = "h"
        model.periodic_boundaries = "i"
        model.external_field = "j"
        model.method_family = "k"
        prop.set_model(model)
        model_db = prop.get_model()
        assert model.spin_mode == model_db.spin_mode
        assert model.basis_set == model_db.basis_set
        assert model.program == model_db.program
        assert model.version == model_db.version
        assert model.solvation == model_db.solvation
        assert model.solvent == model_db.solvent
        assert model.embedding == model_db.embedding
        assert model.periodic_boundaries == model_db.periodic_boundaries
        assert model.external_field == model_db.external_field
        assert model.method_family == model_db.method_family

    def test_model_fails_collection(self):
        prop = db.Property(db.ID())
        model = db.Model("dft", "pbe", "def2-svp")
        self.assertRaises(RuntimeError, lambda: prop.set_model(model))
        self.assertRaises(RuntimeError, lambda: prop.get_model())

    def test_model_fails_id(self):
        coll = self.manager.get_collection("properties")
        prop = db.NumberProperty()
        prop.link(coll)
        model = db.Model("dft", "pbe", "def2-svp")
        self.assertRaises(RuntimeError, lambda: prop.set_model(model))
        self.assertRaises(RuntimeError, lambda: prop.get_model())

    def test_structure(self):
        # Setup
        coll = self.manager.get_collection("properties")

        model = db.Model("dft", "pbe", "def2-svp")
        prop = db.NumberProperty.make("density_matrix", model, 0.0, coll)
        assert prop.has_id()

        # Tests
        sid = db.ID()
        assert not prop.has_structure()
        prop.set_structure(sid)
        assert prop.has_structure()
        assert prop.get_structure() == sid
        prop.clear_structure()
        assert not prop.has_structure()

    def test_structure_fails_collection(self):
        prop = db.Property(db.ID())
        sid = db.ID()
        self.assertRaises(RuntimeError, lambda: prop.set_structure(sid))
        self.assertRaises(RuntimeError, lambda: prop.get_structure())
        self.assertRaises(RuntimeError, lambda: prop.has_structure())
        self.assertRaises(RuntimeError, lambda: prop.clear_structure())

    def test_structure_fails_id(self):
        coll = self.manager.get_collection("properties")
        prop = db.NumberProperty()
        prop.link(coll)
        sid = db.ID()
        self.assertRaises(RuntimeError, lambda: prop.set_structure(sid))
        self.assertRaises(RuntimeError, lambda: prop.get_structure())
        self.assertRaises(RuntimeError, lambda: prop.has_structure())
        self.assertRaises(RuntimeError, lambda: prop.clear_structure())

    def test_calculation(self):
        # Setup
        coll = self.manager.get_collection("properties")

        model = db.Model("dft", "pbe", "def2-svp")
        prop = db.NumberProperty.make("density_matrix", model, 0.0, coll)
        assert prop.has_id()

        # Tests
        sid = db.ID()
        assert not prop.has_calculation()
        prop.set_calculation(sid)
        assert prop.has_calculation()
        assert prop.get_calculation() == sid
        prop.clear_calculation()
        assert not prop.has_calculation()

    def test_calculation_fails_collection(self):
        prop = db.Property(db.ID())
        sid = db.ID()
        self.assertRaises(RuntimeError, lambda: prop.set_calculation(sid))
        self.assertRaises(RuntimeError, lambda: prop.get_calculation())
        self.assertRaises(RuntimeError, lambda: prop.has_calculation())
        self.assertRaises(RuntimeError, lambda: prop.clear_calculation())

    def test_calculation_fails_id(self):
        coll = self.manager.get_collection("properties")
        prop = db.NumberProperty()
        prop.link(coll)
        sid = db.ID()
        self.assertRaises(RuntimeError, lambda: prop.set_calculation(sid))
        self.assertRaises(RuntimeError, lambda: prop.get_calculation())
        self.assertRaises(RuntimeError, lambda: prop.has_calculation())
        self.assertRaises(RuntimeError, lambda: prop.clear_calculation())

    def test_comment(self):
        # Setup
        coll = self.manager.get_collection("properties")
        model = db.Model("dft", "pbe", "def2-svp")
        prop = db.NumberProperty.make("density_matrix", model, 0.0, coll)
        assert prop.has_id()

        # Tests
        assert not prop.has_comment()
        comment = prop.get_comment()
        assert comment == ""

        prop.set_comment("wubba lubba dub dub")
        assert prop.has_comment()
        comment = prop.get_comment()
        assert comment == "wubba lubba dub dub"

        prop.clear_comment()
        assert not prop.has_comment()
        comment = prop.get_comment()
        assert comment == ""

    def test_comment_fails_collection(self):
        prop = db.Property(db.ID())
        self.assertRaises(RuntimeError, lambda: prop.set_comment("test"))
        self.assertRaises(RuntimeError, lambda: prop.get_comment())
        self.assertRaises(RuntimeError, lambda: prop.has_comment())
        self.assertRaises(RuntimeError, lambda: prop.clear_comment())

    def test_comment_fails_id(self):
        coll = self.manager.get_collection("properties")
        prop = db.NumberProperty()
        prop.link(coll)
        self.assertRaises(RuntimeError, lambda: prop.set_comment("test"))
        self.assertRaises(RuntimeError, lambda: prop.get_comment())
        self.assertRaises(RuntimeError, lambda: prop.has_comment())
        self.assertRaises(RuntimeError, lambda: prop.clear_comment())

    def test_get_derived_fails(self):
        # Setup
        coll = self.manager.get_collection("properties")
        model = db.Model("dft", "pbe", "def2-svp")
        tmp = db.NumberProperty.make("density_matrix", model, 5.6, coll)
        prop = db.Property(tmp.get_id(), coll)
        assert prop.has_id()
        num_prop = prop.get_derived()
        assert num_prop.get_data() == 5.6

    def test_get_derived(self):
        # Setup
        coll = self.manager.get_collection("properties")
        model = db.Model("dft", "pbe", "def2-svp")
        tmp = db.NumberProperty.make("density_matrix", model, 5.6, coll)
        prop = db.Property(tmp.get_id())
        self.assertRaises(RuntimeError, lambda: prop.get_derived())
        prop.link(coll)
        assert prop.has_id()
